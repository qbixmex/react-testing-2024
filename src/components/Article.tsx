import { FC } from 'react';

type Props = {
  heading?: string;
  subHeading?: string;
  children: string;
};

const Article: FC<Props> = ({
  heading = 'No Title',
  subHeading = 'No Subheading',
  children,
}) => {
  return (
    <article>
      <hgroup>
        <h1>{ heading }</h1>
        <h2>{ subHeading }</h2>
      </hgroup>
      <main data-testid="main">{children}</main>
    </article>
  );
};
export default Article;
