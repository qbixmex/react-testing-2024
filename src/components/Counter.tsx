import { FC, useState } from 'react';

type Props = {
  initialValue: number;
};

const Counter: FC<Props> = ({ initialValue }) => {
  const [counter, setCounter] = useState(initialValue);

  const handleIncrement = () => {
    setCounter((prevCounter) => prevCounter + 1);
  };

  const handleReset = () => {
    setCounter(initialValue);
  };

  const handleDecrement = () => {
    if (counter === 0) return;
    setCounter((prevCounter) => prevCounter - 1);
  };

  return (
    <>
      <h2>Counter</h2>
      <p aria-label="counter">{ counter }</p>

      <button
        type="button"
        onClick={handleIncrement}
      >
        increment
      </button>

      <button
        type="button"
        onClick={handleReset}
      >
        reset
      </button>

      <button
        type="button"
        onClick={handleDecrement}
      >
        decrement
      </button>
    </>
  );
};

export default Counter;
