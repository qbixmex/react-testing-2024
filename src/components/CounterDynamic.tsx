import { FC } from 'react';

type Props = {
  count: number;
  handleIncrement?: (value: number) => void;
  handleDecrement?: (value: number) => void;
  handleReset?: () => void;
};

const CounterDynamic: FC<Props> = ({
  count,
  handleIncrement,
  handleDecrement,
  handleReset,
}) => {
  return (
    <>
      <h1>Counter Dynamic</h1>
      <h2>{count}</h2>

      {handleIncrement && (
        <button
          type="button"
          onClick={() => handleIncrement(10)}
        >
          increment
        </button>
      )}

      {handleReset && (
        <button type="button" onClick={handleReset}>
          reset
        </button>
      )}

      {handleDecrement && (
        <button
          type="button"
          onClick={() => handleDecrement(5)}
        >
          decrement
        </button>
      )}
    </>
  );
};

export default CounterDynamic;
