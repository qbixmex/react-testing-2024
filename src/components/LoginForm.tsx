import {
  ChangeEvent, FC, useState,
} from 'react';

export type FormData = {
  email: string;
  password: string;
};

const FORM_INITIAL_STATE: FormData = {
  email: '',
  password: '',
};

type Props = {
  onSubmit: (formData: FormData) => void;
};

const Login: FC<Props> = ({ onSubmit }) => {
  const [formData, setFormData] = useState<FormData>(FORM_INITIAL_STATE);

  const onResetForm = () => {
    setFormData(FORM_INITIAL_STATE);
  };

  const onInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    setFormData(prevFormData => {
      return {
        ...prevFormData,
        [event.target.name]: event.target.value,
      };
    });
  };

  return (
    <form
      id="login-form"
      onSubmit={(event) => {
        event.preventDefault();
        onSubmit(formData);
      }}
    >
      <section>
        <label htmlFor="email">Email</label>
        <input
          id="email"
          type="email"
          name="email"
          onChange={onInputChange}
          value={formData.email}
        />
      </section>

      <section>
        <label htmlFor="password">Password</label>
        <input
          id="password"
          type="password"
          name="password"
          onChange={onInputChange}
          value={formData.password}
        />
      </section>

      <section>
        <button
          id="reset-btn"
          type="button"
          onClick={onResetForm}
        >
          Cancel
        </button>

        <button type="submit">Login</button>
      </section>
    </form>
  );
};

export default Login;
