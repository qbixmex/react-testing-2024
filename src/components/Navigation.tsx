import { Link } from 'react-router-dom';
import styles from './Navigation.module.css';

const Navigation = () => {
  return (
    <nav>
      <section className={styles.list} role="list">
        <p role="listitem">
          <Link to="/">Home</Link>
        </p>
        <p role="listitem">
          <Link to="/about">About</Link>
        </p>
      </section>
    </nav>
  );
};

export default Navigation;
