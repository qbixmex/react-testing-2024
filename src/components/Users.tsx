import { useEffect, useState } from 'react';

interface User {
  id: string;
  name: string;
}

const Users = () => {
  const [users, setUsers] = useState<User[]>([]);
  const [errorMessage, setErrorMessage] = useState<string | null>(null);

  useEffect(() => {
    fetch('https://jsonplaceholder.typicode.com/users')
      .then((response) => response.json())
      .then((data) => setUsers(data))
      .catch((error) => {
        setErrorMessage(error.message);
      });
  }, []);

  return (
    <>
      <h1>Users</h1>

      {errorMessage && <p>{errorMessage}</p>}

      {!errorMessage && (
        <ul>
          {users.map(({ id, name }) => (
            <li key={id}>{name}</li>
          ))}
        </ul>
      )}
    </>
  );
};

export default Users;
