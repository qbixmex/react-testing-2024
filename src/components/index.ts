export { default as Navigation } from './Navigation';
export { default as Article } from './Article';
export { default as Counter } from './Counter';
export { default as CounterDynamic } from './CounterDynamic';
export { default as Users } from './Users';
