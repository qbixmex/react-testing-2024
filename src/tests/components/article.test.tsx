import { describe, expect, test } from 'vitest';
import { render } from '@testing-library/react';
import { Article } from '../../components';

describe('Tests on <Article /> component', () => {
  test('Should render with default props', () => {
    const { getByRole, getByText, getByTestId } = render(
      <Article>Lorem ipsum asimet exelsious deu</Article>,
    );

    const heading = getByRole('heading', { level: 1 });
    const subHeading = getByRole('heading', {
      level: 2,
      name: 'No Subheading',
    });
    const text = getByText(/lorem ipsum/i);
    const price = getByTestId('price');

    expect(heading).toBeInTheDocument();
    expect(subHeading).toBeInTheDocument();
    expect(text).toBeInTheDocument();
    expect(price).toBeInTheDocument();
    expect(heading).toHaveTextContent('No Title');
    expect(subHeading).toHaveTextContent('No Subheading');
  });

  test('Should render with provided props', () => {
    const text1 = 'Test Heading';
    const text2 = 'Test Subheading';
    const text3 = 'Test Content';
    const price = 7.95;

    const { getByRole, getByTestId } = render(
      <Article
        heading={text1}
        subHeading={text2}
        price={price}
      >
        {text3}
      </Article>,
    );

    const heading = getByRole('heading', { level: 1 });
    const subHeading = getByRole('heading', {
      level: 2,
      name: text2,
    });
    const main = getByTestId('main');

    expect(heading).toHaveTextContent(text1);
    expect(subHeading).toHaveTextContent(text2);
    expect(main).toHaveTextContent(text3);
  });

  test('Should match with snapshot', () => {
    const text1 = 'Test Heading';
    const text2 = 'Test Subheading';
    const text3 = 'Test Content';
    const price = 32.55;

    const { container } = render(
      <Article
        heading={text1}
        subHeading={text2}
        price={price}
      >
        {text3}
      </Article>,
    );

    expect(container).toMatchSnapshot();
  });

  test('Should contain several html tags', () => {
    const text1 = 'Test Heading';
    const text2 = 'Test Subheading';
    const text3 = 'Test Content';
    const price = 25.95;

    const { container, getByTestId } = render(
      <Article
        heading={text1}
        subHeading={text2}
        price={price}
      >
        {text3}
      </Article>,
    );

    const heading = getByTestId('heading')?.textContent?.trim() as string;
    const subheading = getByTestId('subheading')?.textContent?.trim() as string;
    const paragraph = getByTestId('price')?.innerHTML?.trim() as string;
    const p = container.querySelector('#price-value')?.textContent?.trim() as string;

    expect(heading).toBe(text1);
    expect(subheading).toBe(text2);
    expect(Number(p)).toBe(price);
    expect(paragraph).toContain('$');
    expect(paragraph).toContain(price);
    expect(paragraph).toBe(`$&nbsp;<span id="price-value">${price}</span>`);
    expect(getByTestId('main')).toBeInTheDocument();
  });
});
