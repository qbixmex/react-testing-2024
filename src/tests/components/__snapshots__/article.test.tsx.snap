// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests on <Article /> component > Should match with snapshot 1`] = `
<div>
  <article>
    <hgroup>
      <h1
        data-testid="heading"
      >
        Test Heading
      </h1>
      <h2
        data-testid="subheading"
      >
        Test Subheading
      </h2>
      <p
        data-testid="price"
      >
        $Â 
        <span
          id="price-value"
        >
          32.55
        </span>
      </p>
    </hgroup>
    <main
      data-testid="main"
    >
      Test Content
      <h2>
        Counter
      </h2>
      <p
        aria-label="counter"
      >
        0
      </p>
      <button
        type="button"
      >
        increment
      </button>
      <button
        type="button"
      >
        reset
      </button>
      <button
        type="button"
      >
        decrement
      </button>
    </main>
  </article>
</div>
`;
